File: android/android/src/de/benibela/videlibri/Options.java
package de.benibela.videlibri;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Application;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceCategory;
import android.support.v7.view.ContextThemeWrapper;
import android.util.TypedValue;
import android.view.View;
import android.widget.Checkable;
import android.widget.CompoundButton;
import android.widget.Spinner;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import de.benibela.videlibri.internet.DownloadCertificate;
import de.benibela.videlibri.internet.UserKeyStore;
import de.benibela.videlibri.jni.Bridge;


public class Options extends VideLibriBaseActivity{


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);    //To change body of overridden methods use File | Settings | File Templates.
        syncBridgeToPreferences(this);
        setContentView(R.layout.videlibribaselayout);
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        getSupportFragmentManager().beginTransaction().replace(R.id.content_holder, new SettingsFragment()).commit();
    }

    static public class SettingsFragment extends android.support.v7.preference.PreferenceFragmentCompat  {

        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.preferences);
            Preference[] prefs = new Preference[]{ findPreference("bridge_logging"), findPreference("bridge_nearTime"), findPreference("bridge_refreshInterval") };
            Preference.OnPreferenceChangeListener listener = new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    Bridge.globalOptions = Bridge.VLGetOptions();
                    Bridge.Options options = Bridge.globalOptions;
                    switch (preference.getKey()) {
                        case "bridge_logging":
                            options.logging = (Boolean) newValue;
                            break;
                        case "bridge_nearTime":
                            options.nearTime = (Integer) newValue;
                            break;
                        case "bridge_refreshInterval":
                            options.refreshInterval = (Integer) newValue;
                            break;
                        case "languageOverride":
                            VideLibriApp.setLanguageOverride(getContext(), (String)newValue);
                            //if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
                            //    getActivity().recreate();
                            //}
                    }
                    Bridge.VLSetOptions(options);
                    return true;
                }
            };
            for (Preference p: prefs) p.setOnPreferenceChangeListener(listener);
        }

        @Override
        public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {

        }

        private class CustomPreferenceMaker{
            ContextThemeWrapper contextThemeWrapper;
            PreferenceCategory cat;

            CustomPreferenceMaker(ContextThemeWrapper contextThemeWrapper){
                this.contextThemeWrapper = contextThemeWrapper;
            }

            void beginCat(String key){
                cat = (PreferenceCategory)getPreferenceScreen().findPreference(key);
                cat.removeAll();
            }
            @SuppressWarnings("UnusedReturnValue")
            Preference makePreference(String title, Preference.OnPreferenceClickListener onClick) {
                return makePreference(title, null, onClick);
            }
            Preference makePreference(String title, String summary, Preference.OnPreferenceClickListener onClick) {
                Preference pref = new Preference(contextThemeWrapper);
                pref.setTitle(title);
                if (summary != null) pref.setSummary(summary);
                pref.setOnPreferenceClickListener(onClick);
                cat.addPreference(pref);
                return pref;
            }
        }

        @Override
        public void onResume() {
            super.onResume();
            updatePreferences();
        }

        private void updatePreferences (){
            TypedValue themeTypedValue = new TypedValue();
            getContext().getTheme().resolveAttribute(R.attr.preferenceTheme, themeTypedValue, true);
            @SuppressWarnings("RestrictedApi")
            CustomPreferenceMaker cpm = new CustomPreferenceMaker(new ContextThemeWrapper(getContext(), themeTypedValue.resourceId));

            cpm.beginCat("accounts");

            String summary = getString(R.string.lay_options_label_accounts_summary);
            for (final Bridge.Account acc: VideLibriApp.accounts) if (acc != null) {
                cpm.makePreference(acc.prettyName, summary, new Preference.OnPreferenceClickListener(){
                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        Intent intent = new Intent(getActivity(), AccountInfo.class);
                        intent.putExtra("mode", AccountInfo.MODE_ACCOUNT_MODIFY);
                        intent.putExtra("account", acc);
                        startActivity(intent);
                        return true;
                    }
                });
            }

            cpm.makePreference(getString(R.string.lay_options_btn_newaccount), new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    Intent intent = new Intent(getActivity(), AccountInfo.class);
                    intent.putExtra("mode", AccountInfo.MODE_ACCOUNT_CREATION) ;
                    startActivityForResult(intent, NEW_ACCOUNT_CREATION_RESULT);
                    return true;
                }
            });


            cpm.beginCat("ownlibraries");

            Bridge.Options options = Bridge.VLGetOptions();

            summary = getString(R.string.lay_options_label_ownlibraries_summary);
            for (final String userLibId: options.roUserLibIds) if (userLibId != null) {
                final Bridge.LibraryDetails details = Bridge.VLGetLibraryDetails(userLibId);
                if (details == null) continue;
                cpm.makePreference(details.prettyName, summary, new Preference.OnPreferenceClickListener() {
                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        Intent intent = new Intent(getActivity(), NewLibrary.class);
                        intent.putExtra("mode", NewLibrary.MODE_LIBRARY_MODIFY);
                        intent.putExtra("libId", userLibId);
                        startActivity(intent);
                        return true;
                    }
                });
            }

            cpm.makePreference(getString(R.string.lay_options_btn_newlib), new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    startActivity(new Intent(getActivity(), NewLibrary.class));
                    return true;
                }
            });
            cpm.makePreference(getString(R.string.lay_options_btn_editsource), new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    startActivity(new Intent(getActivity(), SourceEdit.class));
                    return true;
                }
            });

            cpm.beginCat("owncertificates");
            if (UserKeyStore.hasCertificates())
                for (final byte[] cert: UserKeyStore.getCertificates()) {
                    cpm.makePreference(UserKeyStore.getFingerprint(cert), getString(R.string.lay_options_btn_newcertificate_delete), new Preference.OnPreferenceClickListener() {
                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        Bundle b = new Bundle();
                        b.putByteArray("cert", cert);
                        Util.showMessageYesNo(DialogId.OPTIONS_DELETE_USER_CERTIFICATE, getString(R.string.certificate_delete), b);
                        return true;
                    }});
                }

            cpm.makePreference(getString(R.string.lay_options_btn_newcertificate), new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    String defaultServer = "";
                    if (VideLibriApp.errors != null)
                        for (Bridge.PendingException e: VideLibriApp.errors)
                            if (e.kind == Bridge.PendingException.KIND_INTERNET)
                                if (e.error != null && e.error.contains("https://")) {
                                    Matcher matcher = Pattern.compile("https://([^/]+)").matcher(e.error);
                                    if (matcher.find()) {
                                        defaultServer = matcher.group(1);
                                        break;
                                    }
                                }

                    Util.inputDialog(DialogId.OPTIONS_ADD_USER_CERTIFICATE, getString(R.string.certificate_download), null, defaultServer);
                    return true;
                }
            });
        }

        @Override
        public void onDisplayPreferenceDialog(Preference preference) {
            if (preference instanceof PreferenceInteger) {

            } else super.onDisplayPreferenceDialog(preference);
        }
    }


    static final int NEW_ACCOUNT_CREATION_RESULT = 1235;

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == NEW_ACCOUNT_CREATION_RESULT && resultCode == AccountInfo.RESULT_OK)
            finish();
        else super.onActivityResult(requestCode, resultCode, data);
    }



    @Override
    boolean onDialogResult(int dialogId, int buttonId, Bundle more) {
        switch (dialogId) {
            case DialogId.OPTIONS_ADD_USER_CERTIFICATE:
                if (buttonId != Util.MessageHandlerCanceled)
                    (new Thread(new DownloadCertificate(more.getString("text")))).start();
                return true;
            case DialogId.OPTIONS_DELETE_USER_CERTIFICATE:
                if (buttonId == DialogInterface.BUTTON_POSITIVE) {
                    UserKeyStore.removeUserCertificate(more.getByteArray("cert"));
                    UserKeyStore.storeUserCertificates(PreferenceManager.getDefaultSharedPreferences(VideLibriApp.currentContext()));
                    updatePreferences();
                }
                return true;
            default:
                return super.onDialogResult(dialogId, buttonId, more);
        }
    }

    public void updatePreferences() {
        for (Fragment f: getSupportFragmentManager().getFragments())
            if (f instanceof SettingsFragment)
                ((SettingsFragment)f).updatePreferences();
    }

    @Override
    protected void onPause() {
        super.onPause();

        NotificationService.resheduleDailyIfNecessary(this, false);
    }


    static void syncBridgeToPreferences(Activity activity){
        Bridge.globalOptions = Bridge.VLGetOptions();
        Bridge.Options options = Bridge.globalOptions;
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(activity);
        SharedPreferences.Editor editor = sp.edit();
        editor.putBoolean("bridge_logging", options.logging);
        editor.putInt("bridge_nearTime", options.nearTime);
        editor.putInt("bridge_refreshInterval", options.refreshInterval);
        editor.apply();
    }
    /*static void syncPreferencesToBridge(Activity activity){
        Bridge.Options options = Bridge.VLGetOptions();
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(activity);
        options.logging = sp.getBoolean("bridge_logging", false);
        options.nearTime = sp.getInt("bridge_nearTime", 3);
        options.refreshInterval = sp.getInt("bridge_refreshInterval", 1);
        Bridge.VLSetOptions(options);
    }*/

    static void showLendingOptionsInView(Activity activity, View v){
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(activity);
        ((Checkable)v.findViewById(R.id.viewHistory)).setChecked(VideLibri.displayHistory);


        final String[] sortingKeys = activity.getResources().getStringArray(R.array.sortable_properties);
        String sorting = sp.getString("sorting", "dueDate"),
                grouping = sp.getString("grouping", "_dueWeek");
        setSpinnerSelection ((Spinner)v.findViewById(R.id.sorting), sortingKeys, sorting);
        setSpinnerSelection ((Spinner)v.findViewById(R.id.grouping), sortingKeys, grouping);
    }

    @SuppressLint("ApplySharedPref")
    static void putLendingOptionsFromView(Activity activity, View v){
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(activity);
        SharedPreferences.Editor editor = sp.edit();

        VideLibri.displayHistory = ((CompoundButton) v.findViewById(R.id.viewHistory)).isChecked();
        editor.putBoolean("displayHistory", VideLibri.displayHistory);

        final String[] sortingKeys = activity.getResources().getStringArray(R.array.sortable_properties);
        int sortingPos = ((Spinner)v.findViewById(R.id.sorting)).getSelectedItemPosition();
        if (sortingPos >= 0 && sortingPos < sortingKeys.length) editor.putString("sorting", sortingKeys[sortingPos]);
        int groupingPos = ((Spinner)v.findViewById(R.id.grouping)).getSelectedItemPosition();
        if (groupingPos >= 0 && groupingPos < sortingKeys.length) editor.putString("grouping", sortingKeys[groupingPos]);

        editor.commit();
    }

}

File: android/android/src/de/benibela/videlibri/Options.kt
package de.benibela.videlibri

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.preference.PreferenceManager
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.appcompat.view.ContextThemeWrapper
import android.util.TypedValue
import android.view.View
import android.widget.CompoundButton
import android.widget.Spinner
import de.benibela.videlibri.internet.DownloadCertificate
import de.benibela.videlibri.internet.UserKeyStore
import de.benibela.videlibri.jni.Bridge
import de.benibela.videlibri.notifications.NotificationScheduling
import java.util.regex.Pattern


class Options : VideLibriBaseActivity() {


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        syncBridgeToPreferences(this)
        setContentView(R.layout.videlibribaselayout)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        supportFragmentManager.beginTransaction().replace(R.id.content_holder, SettingsFragment()).commit()
    }

    class SettingsFragment : androidx.preference.PreferenceFragmentCompat() {

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            addPreferencesFromResource(R.xml.preferences)
            val prefs = arrayOf(findPreference("bridge_logging"), findPreference("bridge_nearTime"), findPreference("bridge_refreshInterval"))
            val listener = Preference.OnPreferenceChangeListener { preference, newValue ->
                Bridge.globalOptions = Bridge.VLGetOptions()
                val options = Bridge.globalOptions
                when (preference.key) {
                    "bridge_logging" -> options.logging = newValue as Boolean
                    "bridge_nearTime" -> options.nearTime = newValue as Int
                    "bridge_refreshInterval" -> options.refreshInterval = newValue as Int
                    "languageOverride" -> context?.let { VideLibriApp.setLanguageOverride(it, newValue as String) }
                }//if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
                //    getActivity().recreate();
                //}
                Bridge.VLSetOptions(options)
                true
            }
            for (p in prefs) p.onPreferenceChangeListener = listener
        }

        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {

        }

        private inner class CustomPreferenceMaker internal constructor(
                internal var contextThemeWrapper: ContextThemeWrapper
        ) {
            internal lateinit var cat: PreferenceCategory

            internal fun beginCat(key: String) {
                cat = preferenceScreen.findPreference(key) as PreferenceCategory
                cat.removeAll()
            }

            internal fun makePreference(title: String, onClick: Preference.OnPreferenceClickListener): Preference {
                return makePreference(title, null, onClick)
            }

            internal fun makePreference(title: String, summary: String?, onClick: Preference.OnPreferenceClickListener): Preference {
                return Preference(contextThemeWrapper).also {
                    it.title = title
                    if (summary != null) it.summary = summary
                    it.onPreferenceClickListener = onClick
                    cat.addPreference(it)
                }
            }
        }

        override fun onResume() {
            super.onResume()
            updatePreferences()
        }

        internal fun updatePreferences() {
            val context = context ?: return
            val themeTypedValue = TypedValue()
            val theme = context.theme
            theme?.resolveAttribute(R.attr.preferenceTheme, themeTypedValue, true)
            val cpm = CustomPreferenceMaker(ContextThemeWrapper(context, themeTypedValue.resourceId))

            cpm.beginCat("accounts")

            var summary = getString(R.string.lay_options_label_accounts_summary)
            accounts.forEach {acc ->
                cpm.makePreference(acc.prettyName, summary, Preference.OnPreferenceClickListener {
                    startActivity<AccountInfo>(
                            "mode" to AccountInfo.MODE_ACCOUNT_MODIFY,
                            "account" to acc
                    )
                    true
                })
            }

            cpm.makePreference(getString(R.string.lay_options_btn_newaccount), Preference.OnPreferenceClickListener {
                startActivity<AccountInfo>(
                        "mode" to AccountInfo.MODE_ACCOUNT_CREATION
                )
                true
            })


            cpm.beginCat("ownlibraries")

            val options = Bridge.VLGetOptions()

            summary = getString(R.string.lay_options_label_ownlibraries_summary)
            options.roUserLibIds.filterNotNull().forEach { userLibId ->
                val details = Bridge.VLGetLibraryDetails(userLibId) ?: return@forEach
                cpm.makePreference(details.prettyName, summary, Preference.OnPreferenceClickListener {
                    startActivity<NewLibrary>(
                            "mode" to NewLibrary.MODE_LIBRARY_MODIFY,
                            "libId" to userLibId
                    )
                    true
                })
            }

            cpm.makePreference(getString(R.string.lay_options_btn_newlib), Preference.OnPreferenceClickListener {
                startActivity<NewLibrary>()
                true
            })
            cpm.makePreference(getString(R.string.lay_options_btn_editsource), Preference.OnPreferenceClickListener {
                startActivity<SourceEdit>()
                true
            })

            cpm.beginCat("owncertificates")
            if (UserKeyStore.hasCertificates())
                for (cert in UserKeyStore.getCertificates()) {
                    cpm.makePreference(UserKeyStore.getFingerprint(cert), getString(R.string.lay_options_btn_newcertificate_delete), Preference.OnPreferenceClickListener {
                        showMessageYesNo(R.string.certificate_delete) {
                            UserKeyStore.removeUserCertificate(cert)
                            UserKeyStore.storeUserCertificates(PreferenceManager.getDefaultSharedPreferences(VideLibriApp.currentContext()))
                            currentActivity<Options>()?.updatePreferences()
                        }
                        true
                    })
                }

            cpm.makePreference(getString(R.string.lay_options_btn_newcertificate), Preference.OnPreferenceClickListener {
                val defaultServer: String? = VideLibriApp.errors.asSequence().map { e ->
                    if (e.kind == Bridge.PendingException.KIND_INTERNET && e.error?.contains("https://") == true) {
                        val matcher = Pattern.compile("https://([^/]+)").matcher(e.error)
                        if (matcher.find()) matcher.group(1) else null
                    } else null
                }.firstOrNull()

                showInputDialog(R.string.certificate_download, default = defaultServer) { text ->
                    Thread(DownloadCertificate(text)).start()
                }
                true
            })
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == NEW_ACCOUNT_CREATION_RESULT && resultCode == Activity.RESULT_OK)
            finish()
        else
            super.onActivityResult(requestCode, resultCode, data)
    }

    fun updatePreferences() {
        for (f in supportFragmentManager.fragments)
            (f as? SettingsFragment)?.updatePreferences()
    }

    override fun onPause() {
        super.onPause()

        NotificationScheduling.rescheduleDailyIfNecessary(this, false)
    }

    companion object {


        internal const val NEW_ACCOUNT_CREATION_RESULT = 1235


        internal fun syncBridgeToPreferences(activity: Activity) {
            Bridge.globalOptions = Bridge.VLGetOptions()
            val options = Bridge.globalOptions
            PreferenceManager.getDefaultSharedPreferences(activity).edit().apply {
                putBoolean("bridge_logging", options.logging)
                putInt("bridge_nearTime", options.nearTime)
                putInt("bridge_refreshInterval", options.refreshInterval)
                apply()
            }
        }
        /*static void syncPreferencesToBridge(Activity activity){
        Bridge.Options options = Bridge.VLGetOptions();
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(activity);
        options.logging = sp.getBoolean("bridge_logging", false);
        options.nearTime = sp.getInt("bridge_nearTime", 3);
        options.refreshInterval = sp.getInt("bridge_refreshInterval", 1);
        Bridge.VLSetOptions(options);
    }*/

        internal fun showLendingOptionsInView(activity: Activity, v: View) {
            val sp = PreferenceManager.getDefaultSharedPreferences(activity)
            v.findViewById<CompoundButton>(R.id.viewHistory).isChecked = LendingList.displayHistory


            val sortingKeys = activity.resources.getStringArray(R.array.sortable_properties)
            val sorting = sp.getString("sorting", "dueDate")
            val grouping = sp.getString("grouping", "_dueWeek")
            v.findViewById<Spinner>(R.id.sorting).setSelection(sorting, sortingKeys)
            v.findViewById<Spinner>(R.id.grouping).setSelection(grouping, sortingKeys)
        }

        internal fun putLendingOptionsFromView(activity: Activity, v: View) {
            val sortingKeys = activity.resources.getStringArray(R.array.sortable_properties)
            LendingList.displayHistory = v.findViewById<CompoundButton>(R.id.viewHistory).isChecked
            PreferenceManager.getDefaultSharedPreferences(activity).edit().apply {
                putBoolean("displayHistory", LendingList.displayHistory)

                sortingKeys.getOrNull(v.findViewById<Spinner>(R.id.sorting).selectedItemPosition)?.let { putString("sorting", it) }
                sortingKeys.getOrNull(v.findViewById<Spinner>(R.id.grouping).selectedItemPosition)?.let { putString("grouping", it) }

                apply()
            }
        }
    }

}
